### PR TITLE
[DSV3] Optimized Router Gemm

### PR DESCRIPTION
**Pull request by nvmbreughe**
## 📌 Description

This PR:
* adds an optimized router gemm for problem sizes such as Deep Seek-V3. It is ported over from TRTLLM.
* serves as an example on API naming for specialized ops on narrow support surfaces

From my measurements (num tokens = [1,2,4,8,16]), speedups were observed between 1.36 and 1.82x on B200.

Both positive and negative tests were added to test the behavior.

## Breaking Change: Refactored gemm module structure

**ACTION REQUIRED:** Delete stale `flashinfer/gemm.py` file

The `gemm.py` file has been refactored into a package:
- `flashinfer/gemm.py` → `flashinfer/gemm/gemm_base.py`

After pulling this change, run:
```bash
git clean -fd flashinfer/
# OR manually:
rm flashinfer/flashinfer/gemm.py
```

This is backward compatible - no import changes needed.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [ ] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **New Features**
  * High-performance DSv3 router GEMM (bf16 → float32) targeting tokens 1–16, 256 experts, 7168 hidden dim with optional serialized kernel launch.

* **Integration**
  * Python-facing op with runtime shape/dtype/stride validation, backend loader, and registered custom-op entrypoint; public mm_M1_16_K7168_N256 exposed.

* **JIT / Packaging**
  * Adds a JIT module generator and re-exports it for easy import.

* **Tests**
  * Unit tests for positive paths and extensive validation/error cases.

* **Chores**
  * Import-path adjustments and pre-test bytecode cache cleanup.

### CODE REVIEW

File: csrc/dsv3_router_gemm.cu [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+                     __nv_bfloat16 const* weights, cudaStream_t stream) {
+    if (num_tokens == kBegin) {
+      invokeRouterGemm<__nv_bfloat16, kBegin, kNumExperts, kHiddenDim>(output, input, weights,
+                                                                       stream);
+    } else {
+      LoopUnroller<kBegin + 1, kEnd, kNumExperts, kHiddenDim>::unroll(num_tokens, output, input,
+                                                                      weights, stream);
+    }
+  }
+};
```
[the remark this answers is not in the document]
Comment (nvmbreughe) [reply csrc/dsv3_router_gemm.cu:103]: Addressed. Good catch.

**Review by pavanimajety — APPROVED** [review]
Thanks Max! Very thorough PR!

File: scripts/task_test_blackwell_kernels.sh [hunk 2]
```diff
@@ -6,6 +6,13 @@ set -eo pipefail
 : ${MAX_JOBS:=$(nproc)}
 : ${CUDA_VISIBLE_DEVICES:=0}
 
+# Clean Python bytecode cache to avoid stale imports (e.g., after module refactoring)
```
Comment (yzh119) [inline scripts/task_test_blackwell_kernels.sh:9]: Looks like a good practice, I suppose it's mainly useful when we run unittest on the same instance?

File: flashinfer/gemm/routergemm_dsv3.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+import torch
+from flashinfer.utils import (
+    register_custom_op,
+    supported_compute_capability,
+    backend_requirement,
+)
+
+
+# TODO: other compute capabilities may be supported but are untested
+@supported_compute_capability([100])
```
Comment (yzh119) [inline flashinfer/gemm/routergemm_dsv3.py:13]: The codebase doesn't use any features in recent architectures, and the shared memory size usage seems small so I guess it should work for all of our supported architectures (sm75+).

File: flashinfer/gemm/routergemm_dsv3.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    )
+
+
+@backend_requirement({}, common_check=_dsv3_router_gemm_shape_checks)
+def routergemm_dsv3_hidden_7168_experts_256_tokens_lt16(
+    mat_a: torch.Tensor,
+    mat_b: torch.Tensor,
+    out: torch.Tensor,
+    launch_with_pdl: bool = False,
+) -> None:
```
Comment (yzh119) [inline flashinfer/gemm/routergemm_dsv3.py:94]: Can we add a simple docstring here and explicitly mention that it's optimized for small number of tokens.

File: tests/gemm/test_group_gemm.py [hunk 5]
```diff
@@ -33,7 +33,7 @@
     scope="module",
 )
 def warmup_jit():
-    jit_specs = [flashinfer.gemm.gen_gemm_module()]
+    jit_specs = [flashinfer.jit.gemm.gen_gemm_module()]
     if is_sm90a_supported(torch.device("cuda:0")):
         jit_specs.append(flashinfer.gemm.gen_gemm_sm90_module())
```
[the remark this answers is not in the document]
Comment (yzh119) [reply tests/gemm/test_group_gemm.py]: fixed, thanks

**Review by yzh119 — APPROVED** [review]
LGTM

**Review by yongwww — APPROVED** [review]
